### ISSUE TITLE
Output Parquet filenames with uuid to reduce probability of name collisions

### ISSUE THREAD

**Issue by MrPowers** [dask/dask#7657]
[This issue](https://github.com/dask/dask/issues/6867) discusses a related request.

Suppose you write out files like this:

```python
df.to_parquet('../../tmp/people_compressed_parquet', write_index=False, compression='snappy')
```

Here's the files that'll be outputted:

```
part.0.parquet
part.1.parquet
```

If you write again to the same dir, those files will probably get clobbered.

PySpark outputs filenames like the ones below to sidestep the clobbering issue:

```
part-00000-d13891cd-9c59-4441-a078-cd13d0ebbc1d-c000.snappy.parquet
part-00001-d13891cd-9c59-4441-a078-cd13d0ebbc1d-c000.snappy.parquet
```

Adding a full uuid might be a bit excessive.  Perhaps we can output 8 random characters?  I like how Spark adds the compression algo too.  Something like this:

```
part-00000-abcdefgh.snappy.parquet
part-00001-abcdefgh.snappy.parquet
```

Let me know if this suggestion sounds alright and I can try to add a PR.  Thanks!

**Comment by jrbourbeau** [comment]
Thanks @MrPowers. cc @dask/io

**Comment by martindurant** [comment]
> If you write again to the same dir, those files will probably get clobbered.

We do support `append`, the folder should be explicitly cleared before writing if not appending, or the numbering scheme will continue if yes.

The suggestion isn't unreasonable, but I would worry about ending up with mixed datasets. There is no guarantee that the schema of previous leftover files would match the new, and any `_metadata` file might refer only to some set. It seems to me that writing to a temporary directory and *moving* to the final location upon success is the right pattern, but tricky to implement on any filesystem except local.

> I like how Spark adds the compression algo too.

Strongly oppose this - not only is that information not particularly useful (you cannot take a decision based on it), but parquet actually supports different compression per column and different encoding per data page.

**Comment by rjzamora** [comment]
I agree with everything Martin just said, but I also think it is reasonable for the user to have the option to explicitly define a string to be used as the name of every file.  If the user want to include a random uuid and compression information, then so be it.

**Comment by martindurant** [comment]
That's fair enough. fsspec's open_files for writing allows definition of a format string, but this is not currently available via the dask parquet API.

**Comment by MrPowers** [comment]
Thanks for the thoughtful responses and I'll try to give some responses.

I don't feel strongly about any of this and know you're all way more familiar with this codebase and the users.  I'm a Python n00b.  I'm just looking for an excuse to contribute to this code repo 😄 

I did some more experimentation over the weekend and found that CSV files are outputted with no extension, so I'll try to add that via a PR to get started contributing to this project.

> would worry about ending up with mixed datasets. There is no guarantee that the schema of previous leftover files would match the new, and any _metadata file might refer only to some set. It seems to me that writing to a temporary directory and moving to the final location upon success is the right pattern, but tricky to implement on any filesystem except local.

Agree with your concerns, have corrupted data lakes with this very error in the past, and will provide detailed suggestions on a sustainable path forward for this in a future issue.

**Continuing numbering with append**

Does this work with concurrent appends?

> Strongly oppose this (outputting compression algo) - not only is that information not particularly useful (you cannot take a decision based on it)

Is there a good way to determine the compression algo of a file if the file is named `part.0.parquet` and you don't have any additional info to work off of?  [This answer](https://stackoverflow.com/a/19127748) says you can make an educated guess from the first few bytes of the file.  Sorry if I'm missing something obvious here.

> parquet actually supports different compression per column and different encoding per data page.

Didn't realize this, thanks for teaching me!

**Comment by martindurant** [comment]
> Is there a good way to determine the compression algo of a file if the file is named part.0.parquet and you don't have any additional info to work off of? This answer says you can make an educated guess from the first few bytes of the file. Sorry if I'm missing something obvious here.

A "snappy-compressed parquet file" does not mean that you have a normal parquet file and then compress it. Instead, certain data parts within the file are compressed. This is how you have the phenomenally useful feature of parquet to apply different compression algorithms to each column, depending on their data type. Thus, no, you cannot guess the compression from magic bytes, and indeed there is no guarantee of there only being one compression in use. The parquet metadata in the file footer holds all these details.

**Comment by MrPowers** [comment]
@martindurant - I now understand your opposition to including the compression algo in the outputted files and agree with you - thanks for helping me understand.

Looks like the `to_csv` method provides the flexibility to output a UUID in the filename:

```python
import uuid
batch_id = uuid.uuid4()
df.to_csv(f"./tmp/csv_appender/mything-{batch_id}-*.csv", index=False)
```

I like this interface and the flexibility it gives users.  This lets you build a system with concurrent CSV appends without worrying about any filename collisions.

Would you be OK exposing this same interface to `to_parquet`?

```python
df.to_parquet(f"./tmp/parquet_appender/mything-{batch_id}-*.parquet", write_index=False)
```

**Comment by martindurant** [comment]
> Would you be OK exposing this same interface to to_parquet?

In principle yes, but actually parquet writes files into a directory, potentially structured by partitioning. So you wouldn't give an exact glob path like this, but a datafile name template.
